crates/autopilot/src/boundary/events/cow_amm_product_factory.rs

```diff
@@ -0,0 +1,106 @@
+use {
+    contracts::CowAmmConstantProductFactory,
+    ethcontract::{common::DeploymentInformation, dyns::DynWeb3},
+    ethrpc::current_block::RangeInclusive,
+    primitive_types::H160,
+    shared::{event_handling::EventStoring, impl_event_retrieving},
+    std::{
+        collections::{BTreeMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+impl_event_retrieving! {
+    pub CowAmmConstantProductFactoryContract for contracts::cow_amm_constant_product_factory
+}
+
+/// CoW AMM indexer which stores events in-memory.
+#[derive(Clone)]
+pub struct Indexer {
+    /// Registry with the format: (block number, event)
+    registry: Arc<
+        RwLock<BTreeMap<u64, contracts::cow_amm_constant_product_factory::event_data::Deployed>>,
```

Comment:
Since the mutex will be mostly read (for every auction) and barely written, I opted for a `RwLock` which is more efficient in this scenario.